python/lsst/pipe/base/executionButlerBuilder.py

```diff
@@ -130,29 +131,51 @@ def _accumulate(
     datasetTypes: dict[Union[str, DatasetType], DatasetType] = {}
 
     # Add inserts for initOutputs (including initIntermediates); these are
-    # defined fully by their DatasetType, because they have no dimensions, and
-    # they are by definition not resolved.  initInputs are part of Quantum and
-    # that's the only place the graph stores the dataset IDs, so we process
-    # them there even though each Quantum for a task has the same ones.
+    # defined fully by their DatasetType, because they have no dimensions.
+    # initInputs are part of Quantum and that's the only place the graph stores
+    # the dataset IDs, so we process them there even though each Quantum for a
+    # task has the same ones.
     for dataset_type in itertools.chain(dataset_types.initIntermediates, dataset_types.initOutputs):
         dataset_type = _validate_dataset_type(dataset_type, datasetTypes)
         inserts[dataset_type].add(DataCoordinate.makeEmpty(dataset_type.dimensions.universe))
 
+    # Output references may be resolved even if they do not exist. Find all
+    # actually existing refs.
+    check_refs: Set[DatasetRef] = set()
     n: QuantumNode
     for quantum in (n.quantum for n in graph):
         for attrName in ("initInputs", "inputs", "outputs"):
             attr: Mapping[DatasetType, Union[DatasetRef, List[DatasetRef]]] = getattr(quantum, attrName)
-
             for type, refs in attr.items():
                 # This if block is because init inputs has a different
                 # signature for its items
                 if not isinstance(refs, list):
                     refs = [refs]
-                # iterate over all the references, if it has an id, it
-                # means it exists and should be exported, if not it should
-                # be inserted into the new registry
                 for ref in refs:
                     if ref.id is not None:
+                        # We could check existence of individual components,
+                        # but it should be less work to check their parent.
+                        if ref.isComponent():
+                            ref = ref.makeCompositeRef()
+                        check_refs.add(ref)
+    # mexist should be more efficient than checking every ref separately
+    exist_map = butler.datastore.mexists(list(check_refs))
```

Comment:
It will likely be more efficient to call `butler.datastore.knows_these()` since that will do a quick check on datastore registry and won't need to also check that files are there -- previously we assumed that datastore knowing about it is a pretty good hint that the file *should* be there.

Reply:
Thanks for suggestion, I'll try it.

Reply:
Works perfectly, indeed, I need to learn a bit more about that butler thing 🙂